cmd/etcd-launcher/cmd_snapshot.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2023 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"go.etcd.io/etcd/etcdutl/v3/snapshot"
+	"go.uber.org/zap"
+
+	"k8c.io/kubermatic/v2/cmd/etcd-launcher/pkg/etcd"
+)
+
+type snapshotOptions struct {
+	options
+
+	file string
+}
+
+func SnapshotCommand(log *zap.SugaredLogger) *cobra.Command {
+	opt := snapshotOptions{}
+
+	cmd := &cobra.Command{
+		Use:          "snapshot",
+		Short:        "Create etcd database snapshot and save it to file",
+		RunE:         SnapshotFunc(log, &opt),
+		SilenceUsage: true,
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			opts.CopyInto(&opt.options)
+
+			return nil
+		},
+	}
+
+	cmd.SetFlagErrorFunc(func(c *cobra.Command, err error) error {
+		if err := c.Usage(); err != nil {
+			return err
+		}
+
+		// ensure we exit with code 1 later on
+		return err
+	})
+
+	cmd.PersistentFlags().StringVar(&opt.file, "file", "/backup/snapshot.db", "file to save database snapshot to")
+
+	return cmd
+}
+
+func SnapshotFunc(log *zap.SugaredLogger, opt *snapshotOptions) cobraFuncE {
+	return handleErrors(log, func(cmd *cobra.Command, args []string) error {
+		ctx := cmd.Context()
+		log := log.With("cluster", opt.cluster)
+
+		e := &etcd.Cluster{
+			Cluster:           opt.cluster,
+			EtcdctlAPIVersion: opt.etcdctlAPIVersion,
+
+			CaCertFile:     opt.etcdCAFile,
+			ClientCertFile: opt.etcdCertFile,
+			ClientKeyFile:  opt.etcdKeyFile,
+		}
+
+		if _, err := e.Init(ctx); err != nil {
+			return fmt.Errorf("failed to initialize etcd cluster configuration: %w", err)
+		}
+
+		if err := e.SetClusterSize(ctx); err != nil {
+			return fmt.Errorf("failed to set expected cluster size: %w", err)
+		}
+
+		configs, err := e.GetEtcdEndpointConfigs(ctx, log)
+		if err != nil {
+			return fmt.Errorf("failed to get etcd cluster clients: %w", err)
+		}
+
+		for _, config := range configs {
```

Comment:
Can we leave a comment here, indicating that this on purpose tries each endpoint until one successfully created a snapshot? I got confused by why err==nil ends in a return statement, yet errors just result in a loop continuation.